### ISSUE TITLE
Raft consensus network is not operating as it should

### ISSUE THREAD

**Issue by StefPler** [ConsenSys/quorum#1277]
Hello everyone, I have been trying to set up a test network of 3 nodes based on the 7nodes example you have available in github. I am running it without the private configuration, no permissioned nodes and I am using the raft consensus. I have come across something weird and I wanted to see if anyone can help me understand what is going wrong. Also the `static-nodes.json` file is populated in all of the nodes with the information of the cluster. Thanks in advance to anyone who reads this.

# Expected behavior
Nodes are communicating with each other and proper peer number is displayed

# Actual behavior 
Nodes seem to be recognizing one another but the peer count is 0. As a result the network is not functioning properly, meaning I can not connect to it remotely with Metamask for example.

## Outputs
I am attaching some outputs and configuration that might help.
I have removed the actual ip addresses but they are all printed correctly in the actual output.
Inside the nodes when I inspect the raft.cluster object it seems to be working fine.
They all agree on who is the leader and who are the verifies but I am getting no peers from `admin.peers` and 0 peerCount in the `net` object
```
> raft.cluster
[{
    hostname: "ip of node3",
    nodeActive: true,
    nodeId: "579f786d4e2830bbcc02815a27e8a9bacccc9605df4dc6f20bcc1a6eb391e7225fff7cb83e5b4ecd1f3a94d8b733803f2f66b7e871961e7b029e22c155c3a778",
    p2pPort: 21000,
    raftId: 3,
    raftPort: 50400,
    role: "verifier"
}, {
    hostname: "ip of node 2",
    nodeActive: true,
    nodeId: "0ba6b9f606a43a95edc6247cdb1c1e105145817be7bcafd6b2c0ba15d58145f0dc1a194f70ba73cd6f4cdd6864edc7687f311254c7555cc32e4d45aeb1b80416",
    p2pPort: 21000,
    raftId: 2,
    raftPort: 50400,
    role: "verifier"
}, {
    hostname: "ip of node 1",
    nodeActive: true,
    nodeId: "ac6b1096ca56b9f6d004b779ae3728bf83f8e22453404cc3cef16a3d9b96608bc67c4b30db88e0a5a6c6390213f7acbe1153ff6d23ce57380104288ae19373ef",
    p2pPort: 21000,
    raftId: 1,
    raftPort: 50400,
    role: "minter"
}]
```

```
> net
{
  listening: true,
  peerCount: 0,
  version: "10",
  getListening: function(callback),
  getPeerCount: function(callback),
  getVersion: function(callback)
}
> admin.peers
[]
```

The `static-nodes.json` that is used in all of the nodes. Again I have removed the actual ip addresses: 
```
[
 "enode://ac6b1096ca56b9f6d004b779ae3728bf83f8e22453404cc3cef16a3d9b96608bc67c4b30db88e0a5a6c6390213f7acbe1153ff6d23ce57380104288ae19373ef@<node1_ip_address>:21000?discport=0&raftport=50400",
  "enode://0ba6b9f606a43a95edc6247cdb1c1e105145817be7bcafd6b2c0ba15d58145f0dc1a194f70ba73cd6f4cdd6864edc7687f311254c7555cc32e4d45aeb1b80416@<node2_ip_address>:21000?discport=0&raftport=50400",
  "enode://579f786d4e2830bbcc02815a27e8a9bacccc9605df4dc6f20bcc1a6eb391e7225fff7cb83e5b4ecd1f3a94d8b733803f2f66b7e871961e7b029e22c155c3a778@<node3_ip_address>:21000?discport=0&raftport=50400"
]
```

---
I am attaching the docker files I use in case they help. Each one is run on a different machine with different public IPs. As you can see in the docker files I have edited the entrypoints a bit to save the state and removed the `sed` pipe that was overwriting the `static-nodes.json`. 

To run each one of the following docker files I use:

`PRIVATE_CONFIG=ignore QUORUM_CONSENSUS=raft docker-compose up -d`

and I run node 2 and 3 after I have confirmed that node 1 is up.
## Node 1 docker file.

```yaml
# The following environment variables are substituted if present
# * QUORUM_CONSENSUS: default to istanbul
# * QUORUM_DOCKER_IMAGE: default to quorumengineering/quorum:21.7.1
# * QUORUM_TX_MANAGER_DOCKER_IMAGE: default to quorumengineering/tessera:21.7.2
# * QUORUM_GETH_ARGS: extra geth arguments to be included when running geth
# To use Constellation, set QUORUM_TX_MANAGER_DOCKER_IMAGE to Constellation docker image,
# e.g.: QUORUM_TX_MANAGER_DOCKER_IMAGE=quorumengineering/constellation:0.3.2 docker-compose up -d
# To use Remix, set QUORUM_GETH_ARGS="--rpccorsdomain https://remix.ethereum.org"
version: "3.6"
x-quorum-def:
  &quorum-def
  restart: "on-failure"
  image: "${QUORUM_DOCKER_IMAGE:-quorumengineering/quorum:21.7.1}"
  expose:
    - "21000"
    - "50400"
  healthcheck:
    test: ["CMD", "wget", "--spider", "--proxy", "off", "http://localhost:8545"]
    interval: 3s
    timeout: 3s
    retries: 10
    start_period: 5s
  labels:
    com.quorum.consensus: ${QUORUM_CONSENSUS:-istanbul}
  entrypoint:
    - /bin/sh
    - -c
    - |
      UDS_WAIT=10
      if [ "$${PRIVATE_CONFIG}" != "ignore" ]; then
      for i in $$(seq 1 100)
      do
        set -e
        if [ -S $${PRIVATE_CONFIG} ] && \
          [ "I'm up!" == "$$(wget --timeout $${UDS_WAIT} -qO- --proxy off 172.16.239.10$${NODE_ID}:9000/upcheck)" ];
        then break
        else
          echo "Sleep $${UDS_WAIT} seconds. Waiting for TxManager."
          sleep $${UDS_WAIT}
        fi
      done
      fi
      DDIR=/qdata/dd
      GENESIS_FILE="/examples/istanbul-genesis.json"
      if [ "${QUORUM_CONSENSUS:-istanbul}" == "raft" ]; then
        GENESIS_FILE="/examples/genesis.json"
      fi
      NETWORK_ID=$$(cat $${GENESIS_FILE} | grep chainId | awk -F " " '{print $$2}' | awk -F "," '{print $$1}')
      if [ "$$(ls -A $${DDIR})" ]; then
        echo "Chain data found. Restarting existing blockchain.\nIf you wish to reset the bc, stop docker, delete the volumes and docker compose again."
      else
        echo "Creating new chain."
        rm -rf $${DDIR}
        mkdir -p $${DDIR}/keystore
        mkdir -p $${DDIR}/geth
        cp /examples/raft/nodekey$${NODE_ID} $${DDIR}/geth/nodekey
        cp /examples/keys/key$${NODE_ID} $${DDIR}/keystore/
        cp /examples/disallowed-nodes.json $${DDIR}/
        cat /examples/permissioned-nodes.json > $${DDIR}/static-nodes.json
        cp $${DDIR}/static-nodes.json $${DDIR}/permissioned-nodes.json
        cat $${DDIR}/static-nodes.json
        geth --datadir $${DDIR} init $${GENESIS_FILE}
      fi
      GETH_ARGS_raft="--raft --raftport 50400"
      GETH_ARGS_istanbul="--emitcheckpoints --istanbul.blockperiod 1 --mine --minerthreads 1 --syncmode full"
      geth \
        --identity node$${NODE_ID}-${QUORUM_CONSENSUS:-istanbul} \
        --datadir $${DDIR} \
        --nodiscover \
        --verbosity 5 \
        --networkid $${NETWORK_ID} \
        --rpc \
        --rpccorsdomain "*" \
        --rpcvhosts "*" \
        --rpcaddr 0.0.0.0 \
        --rpcport 8545 \
        --rpcapi admin,eth,debug,miner,net,shh,txpool,personal,web3,quorum,$${CONSENSUS_RPC_API} \
        --port 21000 \
        --unlock 0 \
        --allow-insecure-unlock \
        --nousb \
        --password /examples/passwords.txt \
        ${QUORUM_GETH_ARGS:-} $${GETH_ARGS_${QUORUM_CONSENSUS:-istanbul}}
x-cakeshop-def:
  &cakeshop-def
  image: "${CAKESHOP_DOCKER_IMAGE:-quorumengineering/cakeshop:0.12.1}"
  expose:
    - "8999"
  restart: "no"
  healthcheck:
    test: ["CMD", "wget", "--spider", "--proxy=off", "http://localhost:8999/actuator/health"]
    interval: 5s
    timeout: 5s
    retries: 20
    start_period: 5s
  entrypoint:
    - /bin/sh
    - -c
    - |
      DDIR=/qdata/cakeshop/local
      rm -rf $${DDIR}
      mkdir -p $${DDIR}
      cp /examples/cakeshop/application.properties.template $${DDIR}/application.properties
      cp /examples/cakeshop/7nodes_docker.json $${DDIR}/7nodes.json
      java -Xms128M -Xmx128M -Dspring.config.additional-location=file:/qdata/cakeshop/local/ -Dcakeshop.config.dir=/qdata/cakeshop -jar /cakeshop/cakeshop.war
      ;;
services:
  node1:
    << : *quorum-def
    hostname: node1
    ports:
      - "22000:8545"
      - "50400:50400"
    volumes:
      - vol1:/qdata
      - ./examples/7nodes:/examples:ro
    environment:
      - PRIVATE_CONFIG=${PRIVATE_CONFIG:-/qdata/tm/tm.ipc}
      - NODE_ID=1
    networks:
      quorum-examples-net:
        ipv4_address: 172.16.239.11
  cakeshop:
    << : *cakeshop-def
    hostname: cakeshop
    ports:
      - "8999:8999"
    volumes:
      - cakeshopvol:/qdata
      - ./examples/7nodes:/examples:ro
    networks:
      quorum-examples-net:
        ipv4_address: 172.16.239.186
networks:
  quorum-examples-net:
    name: quorum-examples-net
    driver: bridge
    ipam:
      driver: default
      config:
      - subnet: 172.16.239.0/24
volumes:
  "vol1":
  "cakeshopvol":

 ```

## Node 2 docker file 

```yaml
# The following environment variables are substituted if present
# * QUORUM_CONSENSUS: default to istanbul
# * QUORUM_DOCKER_IMAGE: default to quorumengineering/quorum:21.7.1
# * QUORUM_TX_MANAGER_DOCKER_IMAGE: default to quorumengineering/tessera:21.7.2
# * QUORUM_GETH_ARGS: extra geth arguments to be included when running geth
# To use Constellation, set QUORUM_TX_MANAGER_DOCKER_IMAGE to Constellation docker image,
# e.g.: QUORUM_TX_MANAGER_DOCKER_IMAGE=quorumengineering/constellation:0.3.2 docker-compose up -d
# To use Remix, set QUORUM_GETH_ARGS="--rpccorsdomain https://remix.ethereum.org"
version: "3.6"
x-quorum-def:
  &quorum-def
  restart: "on-failure"
  image: "${QUORUM_DOCKER_IMAGE:-quorumengineering/quorum:21.7.1}"
  expose:
    - "21000"
    - "50400"
  healthcheck:
    test: ["CMD", "wget", "--spider", "--proxy", "off", "http://localhost:8545"]
    interval: 3s
    timeout: 3s
    retries: 10
    start_period: 5s
  labels:
    com.quorum.consensus: ${QUORUM_CONSENSUS:-istanbul}
  entrypoint:
    - /bin/sh
    - -c
    - |
      UDS_WAIT=10
      if [ "$${PRIVATE_CONFIG}" != "ignore" ]; then
      for i in $$(seq 1 100)
      do
        set -e
        if [ -S $${PRIVATE_CONFIG} ] && \
          [ "I'm up!" == "$$(wget --timeout $${UDS_WAIT} -qO- --proxy off 172.16.239.10$${NODE_ID}:9000/upcheck)" ];
        then break
        else
          echo "Sleep $${UDS_WAIT} seconds. Waiting for TxManager."
          sleep $${UDS_WAIT}
        fi
      done
      fi
      DDIR=/qdata/dd
      GENESIS_FILE="/examples/istanbul-genesis.json"
      if [ "${QUORUM_CONSENSUS:-istanbul}" == "raft" ]; then
        GENESIS_FILE="/examples/genesis.json"
      fi
      NETWORK_ID=$$(cat $${GENESIS_FILE} | grep chainId | awk -F " " '{print $$2}' | awk -F "," '{print $$1}')
      if [ "$$(ls -A $${DDIR})" ]; then
        echo "Chain data found. Restarting existing blockchain.\nIf you wish to reset the bc, stop docker, delete the volumes and docker compose again."
      else
        echo "Creating new chain."
        rm -rf $${DDIR}
        mkdir -p $${DDIR}/keystore
        mkdir -p $${DDIR}/geth
        cp /examples/raft/nodekey$${NODE_ID} $${DDIR}/geth/nodekey
        cp /examples/keys/key$${NODE_ID} $${DDIR}/keystore/
        cp /examples/disallowed-nodes.json $${DDIR}/
        cat /examples/permissioned-nodes.json > $${DDIR}/static-nodes.json
        cp $${DDIR}/static-nodes.json $${DDIR}/permissioned-nodes.json
        cat $${DDIR}/static-nodes.json
        geth --datadir $${DDIR} init $${GENESIS_FILE}
      fi
      GETH_ARGS_raft="--raft --raftport 50400"
      GETH_ARGS_istanbul="--emitcheckpoints --istanbul.blockperiod 1 --mine --minerthreads 1 --syncmode full"
      geth \
        --identity node$${NODE_ID}-${QUORUM_CONSENSUS:-istanbul} \
        --datadir $${DDIR} \
        --nodiscover \
        --verbosity 5 \
        --networkid $${NETWORK_ID} \
        --rpc \
        --rpccorsdomain "*" \
        --rpcvhosts "*" \
        --rpcaddr 0.0.0.0 \
        --rpcport 8545 \
        --rpcapi admin,eth,debug,miner,net,shh,txpool,personal,web3,quorum,$${CONSENSUS_RPC_API} \
        --raftjoinexisting 2 \
        --port 21000 \
        --unlock 0 \
        --allow-insecure-unlock \
        --nousb \
        --password /examples/passwords.txt \
        ${QUORUM_GETH_ARGS:-} $${GETH_ARGS_${QUORUM_CONSENSUS:-istanbul}}
x-cakeshop-def:
  &cakeshop-def
  image: "${CAKESHOP_DOCKER_IMAGE:-quorumengineering/cakeshop:0.12.1}"
  expose:
    - "8999"
  restart: "no"
  healthcheck:
    test: ["CMD", "wget", "--spider", "--proxy=off", "http://localhost:8999/actuator/health"]
    interval: 5s
    timeout: 5s
    retries: 20
    start_period: 5s
  entrypoint:
    - /bin/sh
    - -c
    - |
      DDIR=/qdata/cakeshop/local
      rm -rf $${DDIR}
      mkdir -p $${DDIR}
      cp /examples/cakeshop/application.properties.template $${DDIR}/application.properties
      cp /examples/cakeshop/7nodes_docker.json $${DDIR}/7nodes.json
      java -Xms128M -Xmx128M -Dspring.config.additional-location=file:/qdata/cakeshop/local/ -Dcakeshop.config.dir=/qdata/cakeshop -jar /cakeshop/cakeshop.war
      ;;
services:
  node2:
    << : *quorum-def
    hostname: node2
    ports:
      - "22000:8545"
      - "50400:50400"
    volumes:
      - vol2:/qdata
      - ./examples/7nodes:/examples:ro
    environment:
      - PRIVATE_CONFIG=${PRIVATE_CONFIG:-/qdata/tm/tm.ipc}
      - NODE_ID=2
    networks:
      quorum-examples-net:
        ipv4_address: 172.16.239.12
  cakeshop:
    << : *cakeshop-def
    hostname: cakeshop
    ports:
      - "8999:8999"
    volumes:
      - cakeshopvol:/qdata
      - ./examples/7nodes:/examples:ro
    networks:
      quorum-examples-net:
        ipv4_address: 172.16.239.186
networks:
  quorum-examples-net:
    name: quorum-examples-net
    driver: bridge
    ipam:
      driver: default
      config:
      - subnet: 172.16.239.0/24
volumes:
  "vol2":
  "cakeshopvol":

```

## Node 3 docker file:
```yaml
# The following environment variables are substituted if present
# * QUORUM_CONSENSUS: default to istanbul
# * QUORUM_DOCKER_IMAGE: default to quorumengineering/quorum:21.7.1
# * QUORUM_TX_MANAGER_DOCKER_IMAGE: default to quorumengineering/tessera:21.7.2
# * QUORUM_GETH_ARGS: extra geth arguments to be included when running geth
# To use Constellation, set QUORUM_TX_MANAGER_DOCKER_IMAGE to Constellation docker image,
# e.g.: QUORUM_TX_MANAGER_DOCKER_IMAGE=quorumengineering/constellation:0.3.2 docker-compose up -d
# To use Remix, set QUORUM_GETH_ARGS="--rpccorsdomain https://remix.ethereum.org"
version: "3.6"
x-quorum-def:
  &quorum-def
  restart: "on-failure"
  image: "${QUORUM_DOCKER_IMAGE:-quorumengineering/quorum:21.7.1}"
  expose:
    - "21000"
    - "50400"
  healthcheck:
    test: ["CMD", "wget", "--spider", "--proxy", "off", "http://localhost:8545"]
    interval: 3s
    timeout: 3s
    retries: 10
    start_period: 5s
  labels:
    com.quorum.consensus: ${QUORUM_CONSENSUS:-istanbul}
  entrypoint:
    - /bin/sh
    - -c
    - |
      UDS_WAIT=10
      if [ "$${PRIVATE_CONFIG}" != "ignore" ]; then
      for i in $$(seq 1 100)
      do
        set -e
        if [ -S $${PRIVATE_CONFIG} ] && \
          [ "I'm up!" == "$$(wget --timeout $${UDS_WAIT} -qO- --proxy off 172.16.239.10$${NODE_ID}:9000/upcheck)" ];
        then break
        else
          echo "Sleep $${UDS_WAIT} seconds. Waiting for TxManager."
          sleep $${UDS_WAIT}
        fi
      done
      fi
      DDIR=/qdata/dd
      GENESIS_FILE="/examples/istanbul-genesis.json"
      if [ "${QUORUM_CONSENSUS:-istanbul}" == "raft" ]; then
        GENESIS_FILE="/examples/genesis.json"
      fi
      NETWORK_ID=$$(cat $${GENESIS_FILE} | grep chainId | awk -F " " '{print $$2}' | awk -F "," '{print $$1}')
      if [ "$$(ls -A $${DDIR})" ]; then
        echo "Chain data found. Restarting existing blockchain.\nIf you wish to reset the bc, stop docker, delete the volumes and docker compose again."
      else
        echo "Creating new chain."
        rm -rf $${DDIR}
        mkdir -p $${DDIR}/keystore
        mkdir -p $${DDIR}/geth
        cp /examples/raft/nodekey$${NODE_ID} $${DDIR}/geth/nodekey
        cp /examples/keys/key$${NODE_ID} $${DDIR}/keystore/
        cp /examples/disallowed-nodes.json $${DDIR}/
        cat /examples/permissioned-nodes.json > $${DDIR}/static-nodes.json
        cp $${DDIR}/static-nodes.json $${DDIR}/permissioned-nodes.json
        cat $${DDIR}/static-nodes.json
        geth --datadir $${DDIR} init $${GENESIS_FILE}
      fi
      GETH_ARGS_raft="--raft --raftport 50400"
      GETH_ARGS_istanbul="--emitcheckpoints --istanbul.blockperiod 1 --mine --minerthreads 1 --syncmode full"
      geth \
        --identity node$${NODE_ID}-${QUORUM_CONSENSUS:-istanbul} \
        --datadir $${DDIR} \
        --nodiscover \
        --verbosity 5 \
        --networkid $${NETWORK_ID} \
        --rpc \
        --rpccorsdomain "*" \
        --rpcvhosts "*" \
        --rpcaddr 0.0.0.0 \
        --rpcport 8545 \
        --rpcapi admin,eth,debug,miner,net,shh,txpool,personal,web3,quorum,$${CONSENSUS_RPC_API} \
        --raftjoinexisting 3 \
        --port 21000 \
        --unlock 0 \
        --allow-insecure-unlock \
        --nousb \
        --password /examples/passwords.txt \
        ${QUORUM_GETH_ARGS:-} $${GETH_ARGS_${QUORUM_CONSENSUS:-istanbul}}
x-cakeshop-def:
  &cakeshop-def
  image: "${CAKESHOP_DOCKER_IMAGE:-quorumengineering/cakeshop:0.12.1}"
  expose:
    - "8999"
  restart: "no"
  healthcheck:
    test: ["CMD", "wget", "--spider", "--proxy=off", "http://localhost:8999/actuator/health"]
    interval: 5s
    timeout: 5s
    retries: 20
    start_period: 5s
  entrypoint:
    - /bin/sh
    - -c
    - |
      DDIR=/qdata/cakeshop/local
      rm -rf $${DDIR}
      mkdir -p $${DDIR}
      cp /examples/cakeshop/application.properties.template $${DDIR}/application.properties
      cp /examples/cakeshop/7nodes_docker.json $${DDIR}/7nodes.json
      java -Xms128M -Xmx128M -Dspring.config.additional-location=file:/qdata/cakeshop/local/ -Dcakeshop.config.dir=/qdata/cakeshop -jar /cakeshop/cakeshop.war
      ;;
services:
  node3:
    << : *quorum-def
    hostname: node3
    ports:
      - "22000:8545"
      - "50400:50400"
    volumes:
      - vol3:/qdata
      - ./examples/7nodes:/examples:ro
    environment:
      - PRIVATE_CONFIG=${PRIVATE_CONFIG:-/qdata/tm/tm.ipc}
      - NODE_ID=3
    networks:
      quorum-examples-net:
        ipv4_address: 172.16.239.13
  cakeshop:
    << : *cakeshop-def
    hostname: cakeshop
    ports:
      - "8999:8999"
    volumes:
      - cakeshopvol:/qdata
      - ./examples/7nodes:/examples:ro
    networks:
      quorum-examples-net:
        ipv4_address: 172.16.239.186
networks:
  quorum-examples-net:
    name: quorum-examples-net
    driver: bridge
    ipam:
      driver: default
      config:
      - subnet: 172.16.239.0/24
volumes:
  "vol3":
  "cakeshopvol":

```

**Comment by baptiste-b-pegasys** [comment]
Hello, I made a test and changed few things.

1. I created a `.env` to put the environment of docker-compose
```
PRIVATE_CONFIG=ignore
QUORUM_CONSENSUS=raft
```

2. I change the port mapping and exposition
```
x-quorum-def:
  &quorum-def
  restart: "on-failure"
  image: "${QUORUM_DOCKER_IMAGE:-quorumengineering/quorum:21.7.1}"
  expose:
    - "21000"
    - "8545"
    - "50400"
 ...
services:
  node1:
    << : *quorum-def
    hostname: node1
    ports:
      - "21000:21000"
      - "8545:8545"
      - "50400:50400"
```

3. I updated the examples/7nodes/permissioned-nodes.json with the list of nodes
```
[
  "enode://ac6b1096ca56b9f6d004b779ae3728bf83f8e22453404cc3cef16a3d9b96608bc67c4b30db88e0a5a6c6390213f7acbe1153ff6d23ce57380104288ae19373ef@192.168.1.41:21000?discport=0&raftport=50400",
  "enode://0ba6b9f606a43a95edc6247cdb1c1e105145817be7bcafd6b2c0ba15d58145f0dc1a194f70ba73cd6f4cdd6864edc7687f311254c7555cc32e4d45aeb1b80416@192.168.1.42:21000?discport=0&raftport=50400",
  "enode://579f786d4e2830bbcc02815a27e8a9bacccc9605df4dc6f20bcc1a6eb391e7225fff7cb83e5b4ecd1f3a94d8b733803f2f66b7e871961e7b029e22c155c3a778@192.168.1.44:21000?discport=0&raftport=50400"
]
```

4. Remove cakeshop in node 2 and node 3

I managed to create the public contract and have the first block mined. The Node 1 still have lot of logs with cakeshop.

I hope this help you to resolve your issue.
Sincerely,

**Comment by StefPler** [comment]
Thank you very much, the issue was the lack of port mapping... My bad. 
Cakeshop seems to not display the other nodes, is there a way to configure it to detect the remote nodes or should I use something else like blockscout for that?
**UPDATE:** I figured out how to do it. Inside the cakeshop folder there is a 7nodes_docker.json, I replaced the rpcUrls there and kept only 3 nodes and it now shows the nodes. However it still says that I have 0 Peers even tho in each node info everyone agrees that peer count is 2.
